diff --git a/Makefile.safaOS b/Makefile.safaOS
new file mode 100644
index 0000000..349d4e7
--- /dev/null
+++ b/Makefile.safaOS
@@ -0,0 +1,245 @@
+# Makefile build
+# meant to be extremely portable to weird unix-like systems
+
+
+CC := clang -target x86_64-unknown-none
+CFLAGS := --sysroot=$(CURDIR)/sysroot -ffreestanding -fno-stack-protector -fno-pie -fno-pic -static -I$(CURDIR)/sysroot/usr/include -nostdlib -D__SAFAOS__ -msse2 -msse3 -DMA_NO_RUNTIME_LINKING -DMA_SUPPORT_SDL \
+      -nostdlibinc
+
+LDFLAGS := --sysroot=$(CURDIR)/sysroot -L$(CURDIR)/sysroot/usr/lib -static -fno-pie -lc -fno-stack-protector -ffreestanding -fno-pic -nostdlib \
+      -nostdlibinc \
+      -Wl,-z,nodlopen
+
+# TODO: Change after porting a good OpenGL SW impl
+DISABLE_MODERN_GL=1
+DISABLE_LEGACY_GL=1
+
+ifeq ($(filter clean distclean,$(MAKECMDGOALS)),)
+
+-include compat/config.mk
+
+ifndef DISABLE_MMD
+DEPFLAGS = -MMD -MP -MF $(@:.$(OBJ_EXT)=.d)
+endif
+
+# trigger configure re-run if $(CC) changes
+_dummy := $(shell \
+	printf '$(CC)' > compat/tmp/cc-new; \
+	cmp -s compat/tmp/cc-new compat/tmp/cc || \
+	mv compat/tmp/cc-new compat/tmp/cc; \
+	rm -f compat/tmp/cc-new \
+)
+
+endif
+
+DEFINES += $(DEFINE)ENABLE_VM_GML_PROFILER \
+		   $(DEFINE)ENABLE_VM_OPCODE_PROFILER \
+		   $(DEFINE)ENABLE_VM_STUB_LOGS \
+		   $(DEFINE)ENABLE_VM_TRACING
+INCLUDES += $(INCLUDE). \
+		    $(INCLUDE)src \
+		    $(INCLUDE)vendor/stb/ds \
+		    $(INCLUDE)src/image \
+		    $(INCLUDE)vendor/stb/image \
+		    $(INCLUDE)vendor/stb/vorbis \
+		    $(INCLUDE)vendor/md5 \
+		    $(INCLUDE)vendor/sha1 \
+		    $(INCLUDE)vendor/base64 \
+		    $(INCLUDE)vendor/bzip2
+
+HEADERS := $(wildcard src/*.h) $(shell find vendor -name '*.h')
+SRCS := $(wildcard src/*.c) $(wildcard src/image/*.c) $(wildcard vendor/bzip2/*.c) vendor/md5/md5.c vendor/sha1/sha1.c vendor/base64/base64.c
+
+DESKTOP_BACKEND := sdl2
+AUDIO_BACKEND := miniaudio
+
+ifdef BUTTERSCOTCH_COMMIT_DATE
+DEFINES += $(DEFINE)BUTTERSCOTCH_COMMIT_DATE=\"$(BUTTERSCOTCH_COMMIT_DATE)\"
+else
+DEFINES += $(DEFINE)BUTTERSCOTCH_COMMIT_DATE=\"unknown\"
+endif
+ifdef BUTTERSCOTCH_COMMIT_HASH
+DEFINES += $(DEFINE)BUTTERSCOTCH_COMMIT_HASH=\"$(BUTTERSCOTCH_COMMIT_HASH)\"
+else
+DEFINES += $(DEFINE)BUTTERSCOTCH_COMMIT_HASH=\"unknown\"
+endif
+
+ifndef DISABLE_WAD14
+DEFINES += $(DEFINE)ENABLE_WAD14
+endif
+
+ifndef DISABLE_WAD16
+DEFINES += $(DEFINE)ENABLE_WAD16
+endif
+
+ifndef DISABLE_WAD17
+DEFINES += $(DEFINE)ENABLE_WAD17
+endif
+
+# TODO: add support for non-desktop backends
+SRCS += $(wildcard src/desktop/*.c) src/desktop/backends/$(DESKTOP_BACKEND).c
+ifeq ($(OS),Windows)
+PKG_CONFIG_FLAGS := --static
+endif
+INCLUDES += $(INCLUDE)src/desktop
+ifeq ($(DESKTOP_BACKEND),glfw3)
+GLFW3_LIBS += $(shell pkg-config $(PKG_CONFIG_FLAGS) --libs glfw3)
+LIBS += $(GLFW3_LIBS)
+DEFINES += $(DEFINE)USE_GLFW3
+ENABLE_GLAD := 1
+ifdef ENABLE_GLES
+DISABLE_SW_RENDERER := 1
+endif
+endif
+ifeq ($(DESKTOP_BACKEND),glfw2)
+GLFW2_LIBS += $(shell pkg-config $(PKG_CONFIG_FLAGS) --libs libglfw)
+LIBS += $(GLFW2_LIBS)
+DEFINES += $(DEFINE)USE_GLFW2
+ENABLE_GLAD := 1
+ifdef ENABLE_GLES
+DISABLE_SW_RENDERER := 1
+endif
+endif
+ifeq ($(DESKTOP_BACKEND),sdl1)
+SDL1_LIBS += $(shell pkg-config $(PKG_CONFIG_FLAGS) --libs sdl)
+LIBS += $(SDL1_LIBS)
+DEFINES += $(DEFINE)USE_SDL1
+endif
+ifeq ($(DESKTOP_BACKEND),sdl2)
+SDL2_LIBS += $(shell sysroot/usr/bin/sdl2-config --libs)
+LIBS := $(SDL2_LIBS)
+DEFINES += $(DEFINE)USE_SDL2
+endif
+ifeq ($(DESKTOP_BACKEND),sdl3)
+SDL3_LIBS += $(shell pkg-config $(PKG_CONFIG_FLAGS) --libs sdl3)
+LIBS += $(SDL3_LIBS)
+DEFINES += $(DEFINE)USE_SDL3
+endif
+
+
+# GNU make doesn't have a way to do OR in conditionals, stupid language for clowns
+ifndef DISABLE_LEGACY_GL
+ENABLE_GL := 1
+endif
+ifndef DISABLE_MODERN_GL
+ENABLE_GL := 1
+endif
+
+ifdef ENABLE_GL
+SRCS += $(wildcard src/gl_common/*.c)
+INCLUDES += $(INCLUDE)src/gl_common $(INCLUDE)src/gl
+HEADERS += $(wildcard src/gl_common/*.h)
+ENABLE_GLAD := 1
+endif
+
+ifndef DISABLE_LEGACY_GL
+DEFINES += $(DEFINE)ENABLE_LEGACY_GL
+SRCS += $(wildcard src/gl_legacy/*.c)
+INCLUDES += $(INCLUDE)src/gl_legacy
+HEADERS += $(wildcard src/gl_legacy/*.h) $(wildcard src/gl/*.h)
+endif
+
+ifndef DISABLE_MODERN_GL
+DEFINES += $(DEFINE)ENABLE_MODERN_GL
+SRCS += $(wildcard src/gl/*.c)
+HEADERS += $(wildcard src/gl/*.h)
+endif
+
+ifndef DISABLE_SW_RENDERER
+DEFINES += -DENABLE_SW_RENDERER
+SRCS += $(wildcard src/sw/*.c)
+HEADERS += $(wildcard src/sw/*.h)
+INCLUDES += -Isrc/sw
+endif
+
+ifdef DISABLE_WAD14
+ifdef DISABLE_WAD16
+ifdef DISABLE_WAD17
+$(error must enable at least 1 bytecode version)
+endif
+endif
+endif
+
+ifdef DISABLE_LEGACY_GL
+ifdef DISABLE_MODERN_GL
+ifdef DISABLE_SW_RENDERER
+$(error must enable at least 1 renderer)
+endif
+endif
+endif
+
+ifeq ($(AUDIO_BACKEND),miniaudio)
+INCLUDES += $(INCLUDE)src/audio/miniaudio $(INCLUDE)vendor/miniaudio
+DEFINES += $(DEFINE)USE_MINIAUDIO
+SRCS += $(wildcard src/audio/miniaudio/*.c)
+HEADERS += $(wildcard src/audio/miniaudio/*.h)
+ifneq ($(OS),Windows)
+LIBS += -pthread
+endif
+endif
+ifeq ($(AUDIO_BACKEND),openal)
+INCLUDES += $(INCLUDE)src/audio/openal
+DEFINES += $(DEFINE)USE_OPENAL
+SRCS += $(wildcard src/audio/openal/*.c)
+HEADERS += $(wildcard src/audio/openal/*.h)
+ifeq ($(OS),Darwin)
+LIBS += -framework OpenAL
+else
+LIBS += -lopenal
+endif
+endif
+
+ifdef ENABLE_GLAD
+SRCS += vendor/glad/src/glad.c
+INCLUDES += $(INCLUDE)vendor/glad/include
+endif
+
+ifeq ($(OS),Windows)
+ifndef MSVC
+LIBS += -static
+LIBS += -lwinmm
+else
+LIBS += winmm.lib
+DEFINES += $(DEFINE)_CRT_SECURE_NO_WARNINGS
+endif
+DEFINES += $(DEFINE)WIN32_LEAN_AND_MEAN
+else
+ifeq ($(OS),Darwin)
+LIBS += -lobjc
+else
+LIBS += -lc sysroot/usr/lib/crt0.o
+endif
+endif
+
+ifndef VERBOSE
+V := @
+endif
+
+OBJS := $(addprefix build/,$(SRCS:.c=.c.$(OBJ_EXT)))
+
+all: build/butterscotch
+
+-include $(OBJS:.$(OBJ_EXT)=.d)
+
+ifeq ($(filter clean distclean,$(MAKECMDGOALS)),)
+
+compat/config.mk: compat/configure.sh compat/tmp/cc
+	@CC="$(CC)" CFLAGS="$(CFLAGS)" LDFLAGS="$(LDFLAGS)" CC="$(CC)" $(SHELL) compat/configure.sh
+
+endif
+
+build/butterscotch: $(OBJS)
+	@{ [ -z "$(NO_COLOR)" ] && [ -t 1 ]; } && printf " \033[1;34mLD\033[0m butterscotch\n" || printf " LD butterscotch\n"
+	$(V)$(_CC) $(LDFLAGS) $(OBJS) $(LIBS) $(EXTRALIBS) $(OUTPUT_EXE)$@
+	@[ -f $@.exe ] && chmod +x $@.exe || true
+
+build/%.c.$(OBJ_EXT): %.c compat/config.mk $(if $(DISABLE_MMD),$(HEADERS))
+	@mkdir -p $(dir $@)
+	@{ [ -z "$(NO_COLOR)" ] && [ -t 1 ]; } && printf " \033[1;32mCC\033[0m $<\n" || printf " CC $<\n"
+	$(V)$(_CC) $(DEFINES) $(INCLUDES) $(CFLAGS) $(DEPFLAGS) $(COMPILE_OBJ) $< $(OUTPUT_OBJ)$@
+
+clean:
+	rm -rf build
+
+distclean: clean
+	rm -f compat/config.mk compat/tmp/cc
diff --git a/compat/configure.sh b/compat/configure.sh
index c549391..509f9db 100644
--- a/compat/configure.sh
+++ b/compat/configure.sh
@@ -62,8 +62,8 @@ check() {
     shift
     output="$output_exe"
     [ -n "$nolink" ] && output="$compile_obj $output_obj" && nolink=
-    printf 'cmd: %s\n' "$CC $cflags tmp/test.c ${output}tmp/a.out $*" >> tmp/config.log
-    if $CC $cflags tmp/test.c ${output}tmp/a.out "$@" >> tmp/config.log 2>&1; then
+    printf 'cmd: %s\n' "$CC $CFLAGS $cflags tmp/test.c $LDFLAGS ${output}tmp/a.out $*" >> tmp/config.log
+    if $CC $CFLAGS $cflags tmp/test.c ${output}tmp/a.out "$@" $LDFLAGS >> tmp/config.log 2>&1; then
         printyes
         return 0
     else
@@ -100,19 +100,18 @@ if $CC /nologo tmp/test.c /Fe:tmp/a.out >> tmp/config.log 2>&1; then
     config 'MSVC := 1'
     config 'OBJ_EXT := obj'
     config "_CC := \$(CC) /nologo"
-    config 'CFLAGS := /O2 /DNDEBUG'
+    config 'CFLAGS += /O2 /DNDEBUG'
     config 'INCLUDE := /I'
     config 'DEFINE := /D'
 elif $CC tmp/test.c -o tmp/a.out >> tmp/config.log 2>&1; then
     printgreen 'gcc'
     syntax=gcc
-    lm='-lm'
     compile_obj='-c'
     output_obj='-o'
     output_exe='-o'
     config 'OBJ_EXT := o'
     config "_CC := \$(CC)"
-    config 'CFLAGS := -O2 -DNDEBUG'
+    config 'CFLAGS += -O2 -DNDEBUG'
     config 'INCLUDE := -I'
     config 'DEFINE := -D'
 else
diff --git a/src/audio/miniaudio/ma_audio_system.c b/src/audio/miniaudio/ma_audio_system.c
index 3ec79ec..307cbdb 100644
--- a/src/audio/miniaudio/ma_audio_system.c
+++ b/src/audio/miniaudio/ma_audio_system.c
@@ -27,6 +27,10 @@
 #include <string.h>
 #include "stb_ds.h"
 
+#ifdef __SAFAOS__
+#include "sdl_backend.h"
+#endif
+
 // ===[ Helpers ]===
 
 static SoundInstance* findFreeSlot(MaAudioSystem* ma) {
@@ -96,7 +100,32 @@ static char* resolveExternalPath(MaAudioSystem* ma, Sound* sound) {
 // ===[ Vtable Implementations ]===
 
 static void maInit(AudioSystem* audio, DataWin* dataWin, FileSystem* fileSystem) {
+#ifdef __SAFAOS__
+    ma_context_config contextConfig;
+
     MaAudioSystem* ma = (MaAudioSystem*) audio;
+
+
+    /*
+    We're just using ma_backend_custom in this example for demonstration purposes, but a more realistic use case would probably want to include
+    other backends as well for robustness.
+    */
+    ma_backend backends[] = {
+        ma_backend_custom
+    };
+
+
+
+    contextConfig = ma_context_config_init();
+    contextConfig.custom.onContextInit = ma_context_init__custom_loader;
+
+    ma_result ctx_result = ma_context_init(backends, sizeof(backends)/sizeof(backends[0]), &contextConfig, (ma_context*)&ma->context);
+    assert(ctx_result == MA_SUCCESS);
+
+    ma_context* finalContext = (ma_context*)&ma->context;
+#else
+    ma_context* finalContext = NULL;
+#endif
     arrput(ma->base.audioGroups, dataWin);
     ma->fileSystem = fileSystem;
 
@@ -107,7 +136,7 @@ static void maInit(AudioSystem* audio, DataWin* dataWin, FileSystem* fileSystem)
     deviceConfig.periodSizeInMilliseconds = 10;
     deviceConfig.dataCallback = ma_engine_data_callback_internal;
     deviceConfig.pUserData    = &ma->engine;
-    ma_result deviceResult = ma_device_init(NULL, &deviceConfig, &ma->device);
+    ma_result deviceResult = ma_device_init(finalContext, &deviceConfig, &ma->device);
     if (deviceResult != MA_SUCCESS) {
         fprintf(stderr, "Audio: Failed to initialize playback device (error %d)\n", deviceResult);
         return;
@@ -167,6 +196,7 @@ static void maDestroy(AudioSystem* audio) {
     arrfree(ma->base.audioGroups);
 
     ma_device_uninit(&ma->device);
+    ma_context_uninit((ma_context*)&ma->context);
     ma_engine_uninit(&ma->engine);
     free(ma);
 }
diff --git a/src/audio/miniaudio/ma_audio_system.h b/src/audio/miniaudio/ma_audio_system.h
index 77f5340..32837fe 100644
--- a/src/audio/miniaudio/ma_audio_system.h
+++ b/src/audio/miniaudio/ma_audio_system.h
@@ -34,8 +34,30 @@ typedef struct {
     float initialPitch;
 } AudioStreamEntry;
 
+#ifdef __SAFAOS__
+typedef struct
+{
+    ma_context context; /* Make this the first member so we can cast between ma_context and ma_context_ex. */
+#if defined(MA_SUPPORT_SDL)
+    struct
+    {
+        ma_handle hSDL; /* A handle to the SDL2 shared object. We dynamically load function pointers at runtime so we can avoid linking. */
+        ma_proc SDL_InitSubSystem;
+        ma_proc SDL_QuitSubSystem;
+        ma_proc SDL_GetNumAudioDevices;
+        ma_proc SDL_GetAudioDeviceName;
+        ma_proc SDL_CloseAudioDevice;
+        ma_proc SDL_OpenAudioDevice;
+        ma_proc SDL_PauseAudioDevice;
+    } sdl;
+#endif
+} ma_context_ex;
+#endif
 typedef struct {
     AudioSystem base;
+    #ifdef __SAFAOS__
+    ma_context_ex context;
+    #endif
     ma_engine engine;
     ma_device device;
     SoundInstance instances[MAX_SOUND_INSTANCES];
diff --git a/src/audio/miniaudio/sdl_backend.h b/src/audio/miniaudio/sdl_backend.h
new file mode 100644
index 0000000..b55b073
--- /dev/null
+++ b/src/audio/miniaudio/sdl_backend.h
@@ -0,0 +1,556 @@
+/* Support SDL on everything. */
+#define MA_SUPPORT_SDL
+
+/*
+Only enable SDL if it's hasn't been explicitly disabled (MA_NO_SDL) or enabled (MA_ENABLE_SDL with
+MA_ENABLE_ONLY_SPECIFIC_BACKENDS) and it's supported at compile time (MA_SUPPORT_SDL).
+*/
+#if defined(MA_SUPPORT_SDL) && !defined(MA_NO_SDL) && (!defined(MA_ENABLE_ONLY_SPECIFIC_BACKENDS) || defined(MA_ENABLE_SDL))
+    #define MA_HAS_SDL
+#endif
+
+
+
+
+typedef struct
+{
+    ma_device device;   /* Make this the first member so we can cast between ma_device and ma_device_ex. */
+#if defined(MA_SUPPORT_SDL)
+    struct
+    {
+        int deviceIDPlayback;
+        int deviceIDCapture;
+    } sdl;
+#endif
+} ma_device_ex;
+
+
+
+#if defined(MA_HAS_SDL)
+    /* SDL headers are necessary if using compile-time linking. */
+    #ifdef MA_NO_RUNTIME_LINKING
+        #ifdef __has_include
+            #ifdef MA_EMSCRIPTEN
+                #if !__has_include(<SDL/SDL_audio.h>)
+                    #undef MA_HAS_SDL
+                #endif
+            #else
+                #if !__has_include(<SDL2/SDL_audio.h>)
+                    #undef MA_HAS_SDL
+                #endif
+            #endif
+        #endif
+    #endif
+#endif
+
+
+#if defined(MA_HAS_SDL)
+#define MA_SDL_INIT_AUDIO                       0x00000010
+#define MA_AUDIO_U8                             0x0008
+#define MA_AUDIO_S16                            0x8010
+#define MA_AUDIO_S32                            0x8020
+#define MA_AUDIO_F32                            0x8120
+#define MA_SDL_AUDIO_ALLOW_FREQUENCY_CHANGE     0x00000001
+#define MA_SDL_AUDIO_ALLOW_FORMAT_CHANGE        0x00000002
+#define MA_SDL_AUDIO_ALLOW_CHANNELS_CHANGE      0x00000004
+#define MA_SDL_AUDIO_ALLOW_ANY_CHANGE           (MA_SDL_AUDIO_ALLOW_FREQUENCY_CHANGE | MA_SDL_AUDIO_ALLOW_FORMAT_CHANGE | MA_SDL_AUDIO_ALLOW_CHANNELS_CHANGE)
+
+/* If we are linking at compile time we'll just #include SDL.h. Otherwise we can just redeclare some stuff to avoid the need for development packages to be installed. */
+#ifdef MA_NO_RUNTIME_LINKING
+    #define SDL_MAIN_HANDLED
+    #ifdef MA_EMSCRIPTEN
+        #include <SDL/SDL.h>
+    #else
+        #include <SDL2/SDL.h>
+    #endif
+
+    typedef SDL_AudioCallback   MA_SDL_AudioCallback;
+    typedef SDL_AudioSpec       MA_SDL_AudioSpec;
+    typedef SDL_AudioFormat     MA_SDL_AudioFormat;
+    typedef SDL_AudioDeviceID   MA_SDL_AudioDeviceID;
+#else
+    typedef void (* MA_SDL_AudioCallback)(void* userdata, ma_uint8* stream, int len);
+    typedef ma_uint16 MA_SDL_AudioFormat;
+    typedef ma_uint32 MA_SDL_AudioDeviceID;
+
+    typedef struct MA_SDL_AudioSpec
+    {
+        int freq;
+        MA_SDL_AudioFormat format;
+        ma_uint8 channels;
+        ma_uint8 silence;
+        ma_uint16 samples;
+        ma_uint16 padding;
+        ma_uint32 size;
+        MA_SDL_AudioCallback callback;
+        void* userdata;
+    } MA_SDL_AudioSpec;
+#endif
+
+typedef int                  (* MA_PFN_SDL_InitSubSystem)(ma_uint32 flags);
+typedef void                 (* MA_PFN_SDL_QuitSubSystem)(ma_uint32 flags);
+typedef int                  (* MA_PFN_SDL_GetNumAudioDevices)(int iscapture);
+typedef const char*          (* MA_PFN_SDL_GetAudioDeviceName)(int index, int iscapture);
+typedef void                 (* MA_PFN_SDL_CloseAudioDevice)(MA_SDL_AudioDeviceID dev);
+typedef MA_SDL_AudioDeviceID (* MA_PFN_SDL_OpenAudioDevice)(const char* device, int iscapture, const MA_SDL_AudioSpec* desired, MA_SDL_AudioSpec* obtained, int allowed_changes);
+typedef void                 (* MA_PFN_SDL_PauseAudioDevice)(MA_SDL_AudioDeviceID dev, int pause_on);
+
+MA_SDL_AudioFormat ma_format_to_sdl(ma_format format)
+{
+    switch (format)
+    {
+    case ma_format_unknown: return 0;
+    case ma_format_u8:      return MA_AUDIO_U8;
+    case ma_format_s16:     return MA_AUDIO_S16;
+    case ma_format_s24:     return MA_AUDIO_S32;  /* Closest match. */
+    case ma_format_s32:     return MA_AUDIO_S32;
+    case ma_format_f32:     return MA_AUDIO_F32;
+    default:                return 0;
+    }
+}
+
+ma_format ma_format_from_sdl(MA_SDL_AudioFormat format)
+{
+    switch (format)
+    {
+        case MA_AUDIO_U8:  return ma_format_u8;
+        case MA_AUDIO_S16: return ma_format_s16;
+        case MA_AUDIO_S32: return ma_format_s32;
+        case MA_AUDIO_F32: return ma_format_f32;
+        default:           return ma_format_unknown;
+    }
+}
+
+static ma_result ma_context_enumerate_devices__sdl(ma_context* pContext, ma_enum_devices_callback_proc callback, void* pUserData)
+{
+    ma_context_ex* pContextEx = (ma_context_ex*)pContext;
+    ma_bool32 isTerminated = MA_FALSE;
+    ma_bool32 cbResult;
+    int iDevice;
+
+    /* Playback */
+    if (!isTerminated) {
+        int deviceCount = ((MA_PFN_SDL_GetNumAudioDevices)pContextEx->sdl.SDL_GetNumAudioDevices)(0);
+        for (iDevice = 0; iDevice < deviceCount; ++iDevice) {
+            ma_device_info deviceInfo;
+            MA_ZERO_OBJECT(&deviceInfo);
+
+            deviceInfo.id.custom.i = iDevice;
+            ma_strncpy_s(deviceInfo.name, sizeof(deviceInfo.name), ((MA_PFN_SDL_GetAudioDeviceName)pContextEx->sdl.SDL_GetAudioDeviceName)(iDevice, 0), (size_t)-1);
+
+            if (iDevice == 0) {
+                deviceInfo.isDefault = MA_TRUE;
+            }
+
+            cbResult = callback(pContext, ma_device_type_playback, &deviceInfo, pUserData);
+            if (cbResult == MA_FALSE) {
+                isTerminated = MA_TRUE;
+                break;
+            }
+        }
+    }
+
+    /* Capture */
+    if (!isTerminated) {
+        int deviceCount = ((MA_PFN_SDL_GetNumAudioDevices)pContextEx->sdl.SDL_GetNumAudioDevices)(1);
+        for (iDevice = 0; iDevice < deviceCount; ++iDevice) {
+            ma_device_info deviceInfo;
+            MA_ZERO_OBJECT(&deviceInfo);
+
+            deviceInfo.id.custom.i = iDevice;
+            ma_strncpy_s(deviceInfo.name, sizeof(deviceInfo.name), ((MA_PFN_SDL_GetAudioDeviceName)pContextEx->sdl.SDL_GetAudioDeviceName)(iDevice, 1), (size_t)-1);
+
+            if (iDevice == 0) {
+                deviceInfo.isDefault = MA_TRUE;
+            }
+
+            cbResult = callback(pContext, ma_device_type_capture, &deviceInfo, pUserData);
+            if (cbResult == MA_FALSE) {
+                isTerminated = MA_TRUE;
+                break;
+            }
+        }
+    }
+
+    return MA_SUCCESS;
+}
+
+static ma_result ma_context_get_device_info__sdl(ma_context* pContext, ma_device_type deviceType, const ma_device_id* pDeviceID, ma_device_info* pDeviceInfo)
+{
+    ma_context_ex* pContextEx = (ma_context_ex*)pContext;
+
+#if !defined(__EMSCRIPTEN__)
+    MA_SDL_AudioSpec desiredSpec;
+    MA_SDL_AudioSpec obtainedSpec;
+    MA_SDL_AudioDeviceID tempDeviceID;
+    const char* pDeviceName;
+#endif
+
+    if (pDeviceID == NULL) {
+        if (deviceType == ma_device_type_playback) {
+            pDeviceInfo->id.custom.i = 0;
+            ma_strncpy_s(pDeviceInfo->name, sizeof(pDeviceInfo->name), MA_DEFAULT_PLAYBACK_DEVICE_NAME, (size_t)-1);
+        } else {
+            pDeviceInfo->id.custom.i = 0;
+            ma_strncpy_s(pDeviceInfo->name, sizeof(pDeviceInfo->name), MA_DEFAULT_CAPTURE_DEVICE_NAME, (size_t)-1);
+        }
+    } else {
+        pDeviceInfo->id.custom.i = pDeviceID->custom.i;
+        ma_strncpy_s(pDeviceInfo->name, sizeof(pDeviceInfo->name), ((MA_PFN_SDL_GetAudioDeviceName)pContextEx->sdl.SDL_GetAudioDeviceName)(pDeviceID->custom.i, (deviceType == ma_device_type_playback) ? 0 : 1), (size_t)-1);
+    }
+
+    if (pDeviceInfo->id.custom.i == 0) {
+        pDeviceInfo->isDefault = MA_TRUE;
+    }
+
+    /*
+    To get an accurate idea on the backend's native format we need to open the device. Not ideal, but it's the only way. An
+    alternative to this is to report all channel counts, sample rates and formats, but that doesn't offer a good representation
+    of the device's _actual_ ideal format.
+
+    Note: With Emscripten, it looks like non-zero values need to be specified for desiredSpec. Whatever is specified in
+    desiredSpec will be used by SDL since it uses it just does its own format conversion internally. Therefore, from what
+    I can tell, there's no real way to know the device's actual format which means I'm just going to fall back to the full
+    range of channels and sample rates on Emscripten builds.
+    */
+#if defined(__EMSCRIPTEN__)
+    /* Good practice to prioritize the best format first so that the application can use the first data format as their chosen one if desired. */
+    pDeviceInfo->nativeDataFormatCount = 3;
+    pDeviceInfo->nativeDataFormats[0].format     = ma_format_s16;
+    pDeviceInfo->nativeDataFormats[0].channels   = 0;   /* All channel counts supported. */
+    pDeviceInfo->nativeDataFormats[0].sampleRate = 0;   /* All sample rates supported. */
+    pDeviceInfo->nativeDataFormats[0].flags      = 0;
+    pDeviceInfo->nativeDataFormats[1].format     = ma_format_s32;
+    pDeviceInfo->nativeDataFormats[1].channels   = 0;   /* All channel counts supported. */
+    pDeviceInfo->nativeDataFormats[1].sampleRate = 0;   /* All sample rates supported. */
+    pDeviceInfo->nativeDataFormats[1].flags      = 0;
+    pDeviceInfo->nativeDataFormats[2].format     = ma_format_u8;
+    pDeviceInfo->nativeDataFormats[2].channels   = 0;   /* All channel counts supported. */
+    pDeviceInfo->nativeDataFormats[2].sampleRate = 0;   /* All sample rates supported. */
+    pDeviceInfo->nativeDataFormats[2].flags      = 0;
+#else
+    MA_ZERO_MEMORY(&desiredSpec, sizeof(desiredSpec));
+
+    pDeviceName = NULL;
+    if (pDeviceID != NULL) {
+        pDeviceName = ((MA_PFN_SDL_GetAudioDeviceName)pContextEx->sdl.SDL_GetAudioDeviceName)(pDeviceID->custom.i, (deviceType == ma_device_type_playback) ? 0 : 1);
+    }
+
+    tempDeviceID = ((MA_PFN_SDL_OpenAudioDevice)pContextEx->sdl.SDL_OpenAudioDevice)(pDeviceName, (deviceType == ma_device_type_playback) ? 0 : 1, &desiredSpec, &obtainedSpec, MA_SDL_AUDIO_ALLOW_ANY_CHANGE);
+    if (tempDeviceID == 0) {
+        ma_log_postf(ma_context_get_log(pContext), MA_LOG_LEVEL_ERROR, "Failed to open SDL device.");
+        return MA_FAILED_TO_OPEN_BACKEND_DEVICE;
+    }
+
+    ((MA_PFN_SDL_CloseAudioDevice)pContextEx->sdl.SDL_CloseAudioDevice)(tempDeviceID);
+
+    /* Only reporting a single native data format. It'll be whatever SDL decides is the best. */
+    pDeviceInfo->nativeDataFormatCount = 1;
+    pDeviceInfo->nativeDataFormats[0].format     = ma_format_from_sdl(obtainedSpec.format);
+    pDeviceInfo->nativeDataFormats[0].channels   = obtainedSpec.channels;
+    pDeviceInfo->nativeDataFormats[0].sampleRate = obtainedSpec.freq;
+    pDeviceInfo->nativeDataFormats[0].flags      = 0;
+
+    /* If miniaudio does not support the format, just use f32 as the native format (SDL will do the necessary conversions for us). */
+    if (pDeviceInfo->nativeDataFormats[0].format == ma_format_unknown) {
+        pDeviceInfo->nativeDataFormats[0].format = ma_format_f32;
+    }
+#endif  /* __EMSCRIPTEN__ */
+
+    return MA_SUCCESS;
+}
+
+
+void ma_audio_callback_capture__sdl(void* pUserData, ma_uint8* pBuffer, int bufferSizeInBytes)
+{
+    ma_device_ex* pDeviceEx = (ma_device_ex*)pUserData;
+
+    ma_device_handle_backend_data_callback((ma_device*)pDeviceEx, NULL, pBuffer, (ma_uint32)bufferSizeInBytes / ma_get_bytes_per_frame(pDeviceEx->device.capture.internalFormat, pDeviceEx->device.capture.internalChannels));
+}
+
+void ma_audio_callback_playback__sdl(void* pUserData, ma_uint8* pBuffer, int bufferSizeInBytes)
+{
+    ma_device_ex* pDeviceEx = (ma_device_ex*)pUserData;
+
+    ma_device_handle_backend_data_callback((ma_device*)pDeviceEx, pBuffer, NULL, (ma_uint32)bufferSizeInBytes / ma_get_bytes_per_frame(pDeviceEx->device.playback.internalFormat, pDeviceEx->device.playback.internalChannels));
+}
+
+static ma_result ma_device_init_internal__sdl(ma_device_ex* pDeviceEx, const ma_device_config* pConfig, ma_device_descriptor* pDescriptor)
+{
+    ma_context_ex* pContextEx = (ma_context_ex*)pDeviceEx->device.pContext;
+    MA_SDL_AudioSpec desiredSpec;
+    MA_SDL_AudioSpec obtainedSpec;
+    const char* pDeviceName;
+    int deviceID;
+
+    /*
+    SDL is a little bit awkward with specifying the buffer size, You need to specify the size of the buffer in frames, but since we may
+    have requested a period size in milliseconds we'll need to convert, which depends on the sample rate. But there's a possibility that
+    the sample rate just set to 0, which indicates that the native sample rate should be used. There's no practical way to calculate this
+    that I can think of right now so I'm just using MA_DEFAULT_SAMPLE_RATE.
+    */
+    if (pDescriptor->sampleRate == 0) {
+        pDescriptor->sampleRate = MA_DEFAULT_SAMPLE_RATE;
+    }
+
+    /*
+    When determining the period size, you need to take defaults into account. This is how the size of the period should be determined.
+
+        1) If periodSizeInFrames is not 0, use periodSizeInFrames; else
+        2) If periodSizeInMilliseconds is not 0, use periodSizeInMilliseconds; else
+        3) If both periodSizeInFrames and periodSizeInMilliseconds is 0, use the backend's default. If the backend does not allow a default
+           buffer size, use a default value of MA_DEFAULT_PERIOD_SIZE_IN_MILLISECONDS_LOW_LATENCY or
+           MA_DEFAULT_PERIOD_SIZE_IN_MILLISECONDS_CONSERVATIVE depending on the value of pConfig->performanceProfile.
+
+    Note that options 2 and 3 require knowledge of the sample rate in order to convert it to a frame count. You should try to keep the
+    calculation of the period size as accurate as possible, but sometimes it's just not practical so just use whatever you can.
+
+    A helper function called ma_calculate_buffer_size_in_frames_from_descriptor() is available to do all of this for you which is what
+    we'll be using here.
+    */
+    pDescriptor->periodSizeInFrames = ma_calculate_buffer_size_in_frames_from_descriptor(pDescriptor, pDescriptor->sampleRate, pConfig->performanceProfile);
+
+    /* SDL wants the buffer size to be a power of 2 for some reason. */
+    if (pDescriptor->periodSizeInFrames > 32768) {
+        pDescriptor->periodSizeInFrames = 32768;
+    } else {
+        pDescriptor->periodSizeInFrames = ma_next_power_of_2(pDescriptor->periodSizeInFrames);
+    }
+
+
+    /* We now have enough information to set up the device. */
+    MA_ZERO_OBJECT(&desiredSpec);
+    desiredSpec.freq     = (int)pDescriptor->sampleRate;
+    desiredSpec.format   = ma_format_to_sdl(pDescriptor->format);
+    desiredSpec.channels = (ma_uint8)pDescriptor->channels;
+    desiredSpec.samples  = (ma_uint16)pDescriptor->periodSizeInFrames;
+    desiredSpec.callback = (pConfig->deviceType == ma_device_type_capture) ? ma_audio_callback_capture__sdl : ma_audio_callback_playback__sdl;
+    desiredSpec.userdata = pDeviceEx;
+
+    /* We'll fall back to f32 if we don't have an appropriate mapping between SDL and miniaudio. */
+    if (desiredSpec.format == 0) {
+        desiredSpec.format = MA_AUDIO_F32;
+    }
+
+    pDeviceName = NULL;
+    if (pDescriptor->pDeviceID != NULL) {
+        pDeviceName = ((MA_PFN_SDL_GetAudioDeviceName)pContextEx->sdl.SDL_GetAudioDeviceName)(pDescriptor->pDeviceID->custom.i, (pConfig->deviceType == ma_device_type_playback) ? 0 : 1);
+    }
+
+    deviceID = ((MA_PFN_SDL_OpenAudioDevice)pContextEx->sdl.SDL_OpenAudioDevice)(pDeviceName, (pConfig->deviceType == ma_device_type_playback) ? 0 : 1, &desiredSpec, &obtainedSpec, MA_SDL_AUDIO_ALLOW_ANY_CHANGE);
+    if (deviceID == 0) {
+        ma_log_postf(ma_device_get_log((ma_device*)pDeviceEx), MA_LOG_LEVEL_ERROR, "Failed to open SDL2 device.");
+        return MA_FAILED_TO_OPEN_BACKEND_DEVICE;
+    }
+
+    if (pConfig->deviceType == ma_device_type_playback) {
+        pDeviceEx->sdl.deviceIDPlayback = deviceID;
+    } else {
+        pDeviceEx->sdl.deviceIDCapture = deviceID;
+    }
+
+    /* The descriptor needs to be updated with our actual settings. */
+    pDescriptor->format             = ma_format_from_sdl(obtainedSpec.format);
+    pDescriptor->channels           = obtainedSpec.channels;
+    pDescriptor->sampleRate         = (ma_uint32)obtainedSpec.freq;
+    ma_channel_map_init_standard(ma_standard_channel_map_default, pDescriptor->channelMap, ma_countof(pDescriptor->channelMap), pDescriptor->channels);
+    pDescriptor->periodSizeInFrames = obtainedSpec.samples;
+    pDescriptor->periodCount        = 1;    /* SDL doesn't use the notion of period counts, so just set to 1. */
+
+    return MA_SUCCESS;
+}
+
+static ma_result ma_device_init__sdl(ma_device* pDevice, const ma_device_config* pConfig, ma_device_descriptor* pDescriptorPlayback, ma_device_descriptor* pDescriptorCapture)
+{
+    ma_device_ex* pDeviceEx = (ma_device_ex*)pDevice;
+    ma_context_ex* pContextEx = (ma_context_ex*)pDevice->pContext;
+    ma_result result;
+
+    /* SDL does not support loopback mode, so must return MA_DEVICE_TYPE_NOT_SUPPORTED if it's requested. */
+    if (pConfig->deviceType == ma_device_type_loopback) {
+        return MA_DEVICE_TYPE_NOT_SUPPORTED;
+    }
+
+    if (pConfig->deviceType == ma_device_type_capture || pConfig->deviceType == ma_device_type_duplex) {
+        result = ma_device_init_internal__sdl(pDeviceEx, pConfig, pDescriptorCapture);
+        if (result != MA_SUCCESS) {
+            return result;
+        }
+    }
+
+    if (pConfig->deviceType == ma_device_type_playback || pConfig->deviceType == ma_device_type_duplex) {
+        result = ma_device_init_internal__sdl(pDeviceEx, pConfig, pDescriptorPlayback);
+        if (result != MA_SUCCESS) {
+            if (pConfig->deviceType == ma_device_type_duplex) {
+                ((MA_PFN_SDL_CloseAudioDevice)pContextEx->sdl.SDL_CloseAudioDevice)(pDeviceEx->sdl.deviceIDCapture);
+            }
+
+            return result;
+        }
+    }
+
+    return MA_SUCCESS;
+}
+
+static ma_result ma_device_uninit__sdl(ma_device* pDevice)
+{
+    ma_device_ex* pDeviceEx = (ma_device_ex*)pDevice;
+    ma_context_ex* pContextEx = (ma_context_ex*)pDevice->pContext;
+
+    if (pDevice->type == ma_device_type_capture || pDevice->type == ma_device_type_duplex) {
+        ((MA_PFN_SDL_CloseAudioDevice)pContextEx->sdl.SDL_CloseAudioDevice)(pDeviceEx->sdl.deviceIDCapture);
+    }
+
+    if (pDevice->type == ma_device_type_playback || pDevice->type == ma_device_type_duplex) {
+        ((MA_PFN_SDL_CloseAudioDevice)pContextEx->sdl.SDL_CloseAudioDevice)(pDeviceEx->sdl.deviceIDCapture);
+    }
+
+    return MA_SUCCESS;
+}
+
+static ma_result ma_device_start__sdl(ma_device* pDevice)
+{
+    ma_device_ex* pDeviceEx = (ma_device_ex*)pDevice;
+    ma_context_ex* pContextEx = (ma_context_ex*)pDevice->pContext;
+
+    if (pDevice->type == ma_device_type_capture || pDevice->type == ma_device_type_duplex) {
+        ((MA_PFN_SDL_PauseAudioDevice)pContextEx->sdl.SDL_PauseAudioDevice)(pDeviceEx->sdl.deviceIDCapture, 0);
+    }
+
+    if (pDevice->type == ma_device_type_playback || pDevice->type == ma_device_type_duplex) {
+        ((MA_PFN_SDL_PauseAudioDevice)pContextEx->sdl.SDL_PauseAudioDevice)(pDeviceEx->sdl.deviceIDPlayback, 0);
+    }
+
+    return MA_SUCCESS;
+}
+
+static ma_result ma_device_stop__sdl(ma_device* pDevice)
+{
+    ma_device_ex* pDeviceEx = (ma_device_ex*)pDevice;
+    ma_context_ex* pContextEx = (ma_context_ex*)pDevice->pContext;
+
+    if (pDevice->type == ma_device_type_capture || pDevice->type == ma_device_type_duplex) {
+        ((MA_PFN_SDL_PauseAudioDevice)pContextEx->sdl.SDL_PauseAudioDevice)(pDeviceEx->sdl.deviceIDCapture, 1);
+    }
+
+    if (pDevice->type == ma_device_type_playback || pDevice->type == ma_device_type_duplex) {
+        ((MA_PFN_SDL_PauseAudioDevice)pContextEx->sdl.SDL_PauseAudioDevice)(pDeviceEx->sdl.deviceIDPlayback, 1);
+    }
+
+    return MA_SUCCESS;
+}
+
+static ma_result ma_context_uninit__sdl(ma_context* pContext)
+{
+    ma_context_ex* pContextEx = (ma_context_ex*)pContext;
+
+    ((MA_PFN_SDL_QuitSubSystem)pContextEx->sdl.SDL_QuitSubSystem)(MA_SDL_INIT_AUDIO);
+
+    /* Close the handle to the SDL shared object last. */
+    ma_dlclose(ma_context_get_log(pContext), pContextEx->sdl.hSDL);
+    pContextEx->sdl.hSDL = NULL;
+
+    return MA_SUCCESS;
+}
+
+static ma_result ma_context_init__sdl(ma_context* pContext, const ma_context_config* pConfig, ma_backend_callbacks* pCallbacks)
+{
+    ma_context_ex* pContextEx = (ma_context_ex*)pContext;
+    int resultSDL;
+
+#ifndef MA_NO_RUNTIME_LINKING
+    /* We'll use a list of possible shared object names for easier extensibility. */
+    size_t iName;
+    const char* pSDLNames[] = {
+#if defined(_WIN32)
+        "SDL2.dll"
+#elif defined(__APPLE__)
+        "SDL2.framework/SDL2"
+#else
+        "libSDL2-2.0.so.0"
+#endif
+    };
+
+    (void)pConfig;
+
+    /* Check if we have SDL2 installed somewhere. If not it's not usable and we need to abort. */
+    for (iName = 0; iName < ma_countof(pSDLNames); iName += 1) {
+        pContextEx->sdl.hSDL = ma_dlopen(ma_context_get_log(pContext), pSDLNames[iName]);
+        if (pContextEx->sdl.hSDL != NULL) {
+            break;
+        }
+    }
+
+    if (pContextEx->sdl.hSDL == NULL) {
+        return MA_NO_BACKEND;   /* SDL2 could not be loaded. */
+    }
+
+    /* Now that we have the handle to the shared object we can go ahead and load some function pointers. */
+    pContextEx->sdl.SDL_InitSubSystem      = ma_dlsym(ma_context_get_log(pContext), pContextEx->sdl.hSDL, "SDL_InitSubSystem");
+    pContextEx->sdl.SDL_QuitSubSystem      = ma_dlsym(ma_context_get_log(pContext), pContextEx->sdl.hSDL, "SDL_QuitSubSystem");
+    pContextEx->sdl.SDL_GetNumAudioDevices = ma_dlsym(ma_context_get_log(pContext), pContextEx->sdl.hSDL, "SDL_GetNumAudioDevices");
+    pContextEx->sdl.SDL_GetAudioDeviceName = ma_dlsym(ma_context_get_log(pContext), pContextEx->sdl.hSDL, "SDL_GetAudioDeviceName");
+    pContextEx->sdl.SDL_CloseAudioDevice   = ma_dlsym(ma_context_get_log(pContext), pContextEx->sdl.hSDL, "SDL_CloseAudioDevice");
+    pContextEx->sdl.SDL_OpenAudioDevice    = ma_dlsym(ma_context_get_log(pContext), pContextEx->sdl.hSDL, "SDL_OpenAudioDevice");
+    pContextEx->sdl.SDL_PauseAudioDevice   = ma_dlsym(ma_context_get_log(pContext), pContextEx->sdl.hSDL, "SDL_PauseAudioDevice");
+#else
+    pContextEx->sdl.SDL_InitSubSystem      = (ma_proc)SDL_InitSubSystem;
+    pContextEx->sdl.SDL_QuitSubSystem      = (ma_proc)SDL_QuitSubSystem;
+    pContextEx->sdl.SDL_GetNumAudioDevices = (ma_proc)SDL_GetNumAudioDevices;
+    pContextEx->sdl.SDL_GetAudioDeviceName = (ma_proc)SDL_GetAudioDeviceName;
+    pContextEx->sdl.SDL_CloseAudioDevice   = (ma_proc)SDL_CloseAudioDevice;
+    pContextEx->sdl.SDL_OpenAudioDevice    = (ma_proc)SDL_OpenAudioDevice;
+    pContextEx->sdl.SDL_PauseAudioDevice   = (ma_proc)SDL_PauseAudioDevice;
+#endif  /* MA_NO_RUNTIME_LINKING */
+
+    resultSDL = ((MA_PFN_SDL_InitSubSystem)pContextEx->sdl.SDL_InitSubSystem)(MA_SDL_INIT_AUDIO);
+    if (resultSDL != 0) {
+        ma_dlclose(ma_context_get_log(pContext), pContextEx->sdl.hSDL);
+        return MA_ERROR;
+    }
+
+    /*
+    The last step is to make sure the callbacks are set properly in `pCallbacks`. Internally, miniaudio will copy these callbacks into the
+    context object and then use them for then on for calling into our custom backend.
+    */
+    pCallbacks->onContextInit             = ma_context_init__sdl;
+    pCallbacks->onContextUninit           = ma_context_uninit__sdl;
+    pCallbacks->onContextEnumerateDevices = ma_context_enumerate_devices__sdl;
+    pCallbacks->onContextGetDeviceInfo    = ma_context_get_device_info__sdl;
+    pCallbacks->onDeviceInit              = ma_device_init__sdl;
+    pCallbacks->onDeviceUninit            = ma_device_uninit__sdl;
+    pCallbacks->onDeviceStart             = ma_device_start__sdl;
+    pCallbacks->onDeviceStop              = ma_device_stop__sdl;
+
+    return MA_SUCCESS;
+}
+#endif  /* MA_HAS_SDL */
+
+
+/*
+This is our custom backend "loader". All this does is attempts to initialize our custom backends in the order they are listed. The first
+one to successfully initialize is the one that's chosen. In this example we're just listing them statically, but you can use whatever logic
+you want to handle backend selection.
+
+This is used as the onContextInit() callback in the context config.
+*/
+static ma_result ma_context_init__custom_loader(ma_context* pContext, const ma_context_config* pConfig, ma_backend_callbacks* pCallbacks)
+{
+    ma_result result = MA_NO_BACKEND;
+
+    /* Silence some unused parameter warnings just in case no custom backends are enabled. */
+    (void)pContext;
+    (void)pCallbacks;
+
+    /* SDL. */
+#if !defined(MA_NO_SDL)
+    if (result != MA_SUCCESS) {
+        result = ma_context_init__sdl(pContext, pConfig, pCallbacks);
+    }
+#endif
+
+    /* ... plug in any other custom backends here ... */
+
+    /* If we have a success result we have initialized a backend. Otherwise we need to tell miniaudio about the error so it can skip over our custom backends. */
+    return result;
+}
diff --git a/src/desktop/main.c b/src/desktop/main.c
index b6f1e60..9b8f069 100644
--- a/src/desktop/main.c
+++ b/src/desktop/main.c
@@ -911,7 +911,11 @@ InputRecording* globalInputRecording = nullptr;
 void __asan_set_death_callback(void (*callback)(void));
 #endif
 
+#ifndef __SAFAOS__
 static volatile sig_atomic_t crashSaveInProgress = 0;
+#else
+static volatile int crashSaveInProgress = 0;
+#endif
 
 static void saveRecordingOnCrash(void) {
     if (crashSaveInProgress) return;
@@ -923,14 +927,17 @@ static void saveRecordingOnCrash(void) {
 
 static void crashSignalHandler(int sig) {
     saveRecordingOnCrash();
+#ifndef __SAFAOS__
     signal(sig, SIG_DFL);
     raise(sig);
+#endif
 }
 
 static void installCrashHandlers(void) {
 #if BUTTERSCOTCH_HAS_ASAN
     __asan_set_death_callback(saveRecordingOnCrash);
 #endif
+#ifndef __SAFAOS__
     signal(SIGSEGV, crashSignalHandler);
     signal(SIGABRT, crashSignalHandler);
 #ifdef SIGBUS
@@ -938,6 +945,7 @@ static void installCrashHandlers(void) {
 #endif
     signal(SIGFPE,  crashSignalHandler);
     signal(SIGILL,  crashSignalHandler);
+#endif
 }
 
 void saveInputRecording() {
@@ -951,7 +959,8 @@ void saveInputRecording() {
     }
 }
 
-#ifndef _WIN32
+
+#if !defined(_WIN32) && !defined(__SAFAOS__)
 typedef struct { int key; struct sigaction value; } PreviousSignalActionEntry;
 static PreviousSignalActionEntry* previousSignalActions = nullptr;
 
@@ -1449,7 +1458,7 @@ int main(int argc, char* argv[]) {
         runner->vmContext->alwaysLogStubbedFunctions = args.alwaysLogStubbedFunctions;
         runner->vmContext->traceEventInherited = args.traceEventInherited;
 
-#ifndef _WIN32
+#if !defined(_WIN32) && !defined(__SAFAOS__)
         struct sigaction sa = {0};
         sa.sa_handler = onCrashSignal;
         sigemptyset(&sa.sa_mask);
@@ -1636,8 +1645,11 @@ int main(int argc, char* argv[]) {
                     memset(runner->keyboard->keyReleased, 0, sizeof(runner->keyboard->keyReleased));
                 }
 
+                uint64_t now = nowNanos();
                 // Run one game step (Begin Step, Keyboard, Alarms, Step, End Step, room transitions)
                 Runner_step(runner);
+                uint64_t diff = nowNanos() - now;
+
 
                 if (freeCamActive) {
                     memcpy(runner->keyboard->keyDown, currentKeyDown, sizeof(runner->keyboard->keyDown));
diff --git a/vendor/miniaudio/miniaudio.h b/vendor/miniaudio/miniaudio.h
index 5055a95..f0390a7 100644
--- a/vendor/miniaudio/miniaudio.h
+++ b/vendor/miniaudio/miniaudio.h
@@ -12006,7 +12006,11 @@ static void ma_sleep__posix(ma_uint32 milliseconds)
     (void)milliseconds;
     MA_ASSERT(MA_FALSE);  /* The Emscripten build should never sleep. */
 #else
-    #if (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 199309L) || defined(MA_SWITCH)
+    #if defined(__SAFAOS__)
+    void SDL_Delay(ma_uint32 ms);
+    // FIXME: Hack, implement nanosleep.
+    SDL_Delay(milliseconds);
+    #elif (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 199309L) || defined(MA_SWITCH)
         struct timespec ts;
         ts.tv_sec  = milliseconds / 1000;
         ts.tv_nsec = milliseconds % 1000 * 1000000;
@@ -13358,7 +13362,7 @@ MA_API ma_result ma_wfopen(FILE** ppFile, const wchar_t* pFilePath, const wchar_
 
         (void)pAllocationCallbacks;
     }
-    #elif !defined(MA_XBOX_NXDK) && !defined(MA_DOS)    /* If your compiler does not support wcsrtombs(), add it here. */
+    #elif !defined(MA_XBOX_NXDK) && !defined(MA_DOS) && !defined(__SAFAOS__)    /* If your compiler does not support wcsrtombs(), add it here. */
     {
         /*
         Use fopen() on anything other than Windows. Requires a conversion. This is annoying because fopen() is locale specific. The only real way I can
@@ -17622,8 +17626,9 @@ static ma_result ma_thread_create__posix(ma_thread* pThread, ma_thread_priority
             (void)stackSize;  /* Suppress unused parameter warning. */
         }
         #endif
-        
 
+
+        #if !defined(__SAFAOS__)
         if (scheduler != -1) {
             int priorityMin = sched_get_priority_min(scheduler);
             int priorityMax = sched_get_priority_max(scheduler);
@@ -17664,6 +17669,9 @@ static ma_result ma_thread_create__posix(ma_thread* pThread, ma_thread_priority
                 }
             }
         }
+        #else
+        (void)priority;
+        #endif
     }
 #else
     /* It's the emscripten build. We'll have a few unused parameters. */
@@ -23061,7 +23069,7 @@ static ma_result ma_context_get_MMDevice__wasapi(ma_context* pContext, ma_device
     CoInitializeResult = ma_CoInitializeEx(pContext, NULL, MA_COINIT_VALUE);
     {
         hr = ma_CoCreateInstance(pContext, &MA_CLSID_MMDeviceEnumerator, NULL, CLSCTX_ALL, &MA_IID_IMMDeviceEnumerator, (void**)&pDeviceEnumerator);
-    }    
+    }
     if (CoInitializeResult == S_OK || CoInitializeResult == S_FALSE) { ma_CoUninitialize(pContext); }
 
     if (FAILED(hr)) {   /* <-- This is checking the call above to ma_CoCreateInstance(). */
@@ -29691,7 +29699,7 @@ static ma_result ma_device_start__alsa(ma_device* pDevice)
     }
 
     if (pDevice->type == ma_device_type_playback || pDevice->type == ma_device_type_duplex) {
-        /*        
+        /*
         When data is written to the device we wait for the device to get ready to receive data with poll(). In my testing
         I have observed that poll() can sometimes block forever unless the device is started explicitly with snd_pcm_start()
         or some data is written with snd_pcm_writei().
@@ -36000,7 +36008,7 @@ static ma_result ma_device_init_internal__coreaudio(ma_context* pContext, ma_dev
             #endif
         }
 
-        
+
         status = ((ma_AudioUnitSetProperty_proc)pContext->coreaudio.AudioUnitSetProperty)(pData->audioUnit, kAudioUnitProperty_StreamFormat, formatScope, formatElement, &bestFormat, sizeof(bestFormat));
         if (status != noErr) {
             ((ma_AudioComponentInstanceDispose_proc)pContext->coreaudio.AudioComponentInstanceDispose)(pData->audioUnit);
@@ -39314,7 +39322,7 @@ static void ma_stream_error_callback__aaudio(ma_AAudioStream* pStream, void* pUs
 
     (void)error;
     ma_log_postf(ma_device_get_log(pDevice), MA_LOG_LEVEL_INFO, "[AAudio] ERROR CALLBACK: error=%d, AAudioStream_getState()=%d\n", error, ((MA_PFN_AAudioStream_getState)pDevice->pContext->aaudio.AAudioStream_getState)(pStream));
-    
+
     /*
     When we get an error, we'll assume that the stream is in an erroneous state and needs to be restarted. From the documentation,
     we cannot do this from the error callback. Therefore we are going to use an event thread for the AAudio backend to do this
@@ -39326,13 +39334,13 @@ static void ma_stream_error_callback__aaudio(ma_AAudioStream* pStream, void* pUs
     else {
         job = ma_job_init(MA_JOB_TYPE_DEVICE_AAUDIO_REROUTE);
         job.data.device.aaudio.reroute.pDevice = pDevice;
-    
+
         if (pStream == pDevice->aaudio.pStreamCapture) {
             job.data.device.aaudio.reroute.deviceType = ma_device_type_capture;
         } else {
             job.data.device.aaudio.reroute.deviceType = ma_device_type_playback;
         }
-    
+
         result = ma_device_job_thread_post(&pDevice->pContext->aaudio.jobThread, &job);
         if (result != MA_SUCCESS) {
             ma_log_postf(ma_device_get_log(pDevice), MA_LOG_LEVEL_INFO, "[AAudio] Device Disconnected. Failed to post job for rerouting.\n");
@@ -39929,7 +39937,7 @@ static ma_result ma_device_reinit__aaudio(ma_device* pDevice, ma_device_type dev
 
     /* We got disconnected! Retry a few times, until we find a connected device! */
     iAttempt = 0;
-    while (iAttempt++ < maxAttempts) {        
+    while (iAttempt++ < maxAttempts) {
         /* Device tearing down? No need to reroute! */
         if (ma_atomic_bool32_get(&pDevice->aaudio.isTearingDown)) {
             result = MA_SUCCESS; /* Caller should continue as normal. */
@@ -40027,7 +40035,7 @@ static ma_result ma_device_reinit__aaudio(ma_device* pDevice, ma_device_type dev
             break;
         }
     }
-    
+
     return result;
 }
 
@@ -61731,7 +61739,7 @@ static ma_result ma_default_vfs_info(ma_vfs* pVFS, ma_vfs_file file, ma_file_inf
         /* Not implemented. Fall back to seek/tell/seek. */
         ma_int64 cursor;
         ma_int64 sizeInBytes;
-        
+
         result = ma_default_vfs_tell(pVFS, file, &cursor);
         if (result != MA_SUCCESS) {
             return result;
@@ -66746,7 +66754,7 @@ static ma_bool32 ma_path_extension_equal_w(const wchar_t* path, const wchar_t* e
     {
         return _wcsicmp(ext1, ext2) == 0;
     }
-    #elif !defined(MA_XBOX_NXDK) && !defined(MA_DOS)
+    #elif !defined(MA_XBOX_NXDK) && !defined(MA_DOS) && !defined(__SAFAOS__)
     {
         /*
         I'm not aware of a wide character version of strcasecmp(). I'm therefore converting the extensions to multibyte strings and comparing those. This
@@ -76964,7 +76972,7 @@ static void ma_engine_node_process_pcm_frames__sound(ma_node* pNode, const float
                 if (pSound->processingCacheFramesRemaining > 0) {
                     MA_MOVE_MEMORY(pSound->pProcessingCache, ma_offset_pcm_frames_ptr_f32(pSound->pProcessingCache, frameCountIn, dataSourceChannels), pSound->processingCacheFramesRemaining * ma_get_bytes_per_frame(ma_format_f32, dataSourceChannels));
                 }
-                
+
                 totalFramesRead += (ma_uint32)frameCountOut;   /* Safe cast. */
 
                 if (result != MA_SUCCESS || ma_sound_at_end(pSound)) {
@@ -78466,7 +78474,7 @@ static ma_result ma_sound_init_from_data_source_internal(ma_engine* pEngine, con
         if (pSound->processingCacheCap == 0) {
             pSound->processingCacheCap = 512;
         }
-        
+
         pSound->pProcessingCache = (float*)ma_calloc(pSound->processingCacheCap * ma_get_bytes_per_frame(ma_format_f32, engineNodeConfig.channelsIn), &pEngine->allocationCallbacks);
         if (pSound->pProcessingCache == NULL) {
             ma_engine_node_uninit(&pSound->engineNode, &pEngine->allocationCallbacks);
diff --git a/vendor/stb/vorbis/stb_vorbis.c b/vendor/stb/vorbis/stb_vorbis.c
index 2f47b56..9fec5ed 100644
--- a/vendor/stb/vorbis/stb_vorbis.c
+++ b/vendor/stb/vorbis/stb_vorbis.c
@@ -588,7 +588,7 @@ enum STBVorbisError
    #if defined(_MSC_VER) || defined(__MINGW32__)
       #include <malloc.h>
    #endif
-   #if defined(__linux__) || defined(__linux) || defined(__sun__) || defined(__EMSCRIPTEN__) || defined(__NEWLIB__)
+   #if defined(__linux__) || defined(__linux) || defined(__sun__) || defined(__EMSCRIPTEN__) || defined(__NEWLIB__) || defined (__SAFAOS__)
       #include <alloca.h>
    #endif
 #else // STB_VORBIS_NO_CRT
